### DISCUSSION TITLE
Issues with RealTimeSTT?

### DISCUSSION THREAD

**Discussion by holdengand** [kimjammer/Neuro#17] [Q&A] [answered]
Are you still providing any kind of support for this project?

I am trying to put this together and get it running in VS Code. I spent a day getting DeepSpeed to install before realizing I was using the wrong python version for my environment for the [deepspeed 0.15.0 version](https://pypi.org/project/RealtimeSTT/0.1.5/#files).

But now when trying to run the main.py I am getting an issue with CoquiEngine hanging up because of issues with RealTimeSTT.

I'm not really a programmer or developer so I'm struggling to piece together what needs to be done or what I have overlooked. But I am always struggling getting the dependencies for RealTimeSTT in order. It always needs a torch version higher or lower then what everything else needs.

I'm not sure its worth even listing what all I have tired as I'm not really even sure I am taking the right approach. I will say I started with installing the requierments and pipfreeze; but got hung up on deepspeed as the version this project used was being a real hassle to get installed on windows; but the 0.15.0 version has sort of resolved the extra things that used to be needed, like doing a WSL install. Swapping the virtual environment to python3.11 made it very easy to get DeepSpeed 0.15.0 installed. But now I can't get past this STT/Torch compatibility thing.

I think I have hit my current limit on what I can do to diagnose this without a smarter mind helping me out. This project seems like the best of its type to me at the moment with the base features I'd like without a bunch of excessive stuff just bloating it up. So its a shame its not getting a bit more work currently. I hope someone can help me figure this out, or someone more capable then me can put together an update.

Here is my current pip freeze 
[pipfreeze.txt](https://github.com/user-attachments/files/17693263/pipfreeze.txt)

**Comment by kimjammer** [comment]
Hey there! Thanks for your interest.
The most likely problem is that you're not using the right deepspeed wheel.

> DeepSpeed (For TTS) will need to be installed separately. I was using instructions from [AllTalkTTS](https://github.com/erew123/alltalk_tts?#-deepspeed-installation-options), and using their [provided wheels](https://github.com/erew123/alltalk_tts/releases/tag/DeepSpeed-14.0).

Click on the "provided wheels" link, and it'll bring you to a place to download the DeepSpeed 0.14 version I was using. If you have further problems, lmk! I don't have as much free time as I did in the summer, but I'll try my best to help.

**Reply by holdengand** [reply]
Right, this is what I was mentioning. I did elect to change which which version of DeepSpeed I was using. Somehow I stupidly got confused and thought you meant the wheel released by DeepSpeed. I didn't even think to check alltalk_tts for it despite it being right there in your readme. Idk what happened there. I promise I was paying attention to your provided instructions early on, but somewhere in the half a dozen reinstalls I must have changed what I was doing.

-Installed the DeepSpeed wheel version suggested from AllTalk
-Reinstalled requirements.txt
-Reinstalled the original pipfreeze.txt

And now I am having issues with the torch version again; the original one I was having before I started looking for other DeepSpeed versions ending up on the 0.15.0 one form DeepSpeed and doing a lot of other dependency shenanigans. This is all in a venv that has only been used to get dependencies for this project. After reinstalling I'm still hung up on RealtimeSTT which wants a higher version of torch, but I can't do that because of DeepSpeed, and probably a few other things.


```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
realtimestt 0.1.16 requires torch==2.3.0, but you have torch 2.2.2+cu118 which is incompatible.
realtimestt 0.1.16 requires torchaudio==2.3.0, but you have torchaudio 2.2.2+cu118 which is incompatible.
```

And running the main.py ends with this issue with CoquiEngine
```
Traceback (most recent call last):
  File "main.py", line 115, in <module>
    asyncio.run(main())
  File "Python\Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "Python\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Python\Python311\Lib\asyncio\base_events.py", line 650, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "main.py", line 50, in main
    tts = TTS(signals)
          ^^^^^^^^^^^^
  File "tts.py", line 13, in __init__
    engine = CoquiEngine(
             ^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable
```

I do have CUDA 11.8 installed
Python 3.11
deepspeed 0.14.0+ce78a63
faster-whisper 1.0.2
pip 24.3.1
RealTimeSTT 0.1.16
RealTimeTTS 0.4.10
torch 2.2.2+cu118
torchaudio 2.2.2+cu118
torchvision 0.17.2+cu118


**It's still probably something I am messing up, I accept that. But left to my own devices I'm just going in circles. Regardless of if you are able to help I appreciate you responding at all honestly.** In my experiences I usually get ignored on git. And I can appreciate you may not have the time, I understand, I am not necessarily in a rush as I don't even have a graphics card to make the best use of this yet. But I was just researching into the project to prep for an upgrade I hope to do in the next few months.

To be clear on one final thing, I am not using oobabooga at this point in time as it was causing me headaches. I am using kobold as it was just quicker to just set up instantly and I know it is also meant to be compatable with Alltalk and xtts. I did change the LLM_ENDPOINT to kobold's local IP and port (5001). I have typically been using Ollama up until messing with this project. I have tested Kobold itself, along with Alltalk plugged in, independently and it works on its own fine. If this is causing the issue, I'll feel dumb but it would be good to clear it up.

**Reply by kimjammer** [reply]
I have good news (hopefully)! I just released v0.2.1 that just updates all the dependencies (including realtimestt and realtimetts), and one of the recent changes to realtimestt is that they [removed the torch version restrictions](https://github.com/KoljaB/RealtimeSTT/commit/1be0a14d139637c663e4319a93d2e3d51c991d44)! Hopefully, if you pull the new version and do a clean installation of Neuro now, you shouldn't run into this problem. BTW, you shouldn't need to install pipfreeze, it's just there to help you troubleshoot, all you should need to do is install requirements, check that torch didn't get overriden, then install deepspeed.

**Reply by kimjammer** [reply]
It shouldn't matter which LLM UI you use, as long as the API is using the OpenAI API specs.

**Reply by holdengand** [reply]
I really really appreciate you taking the effort to update the dependencies. I am having a weird issue though.

Running the requirements, it gets to numba and seems to be complaining that my python version is 3.11.0 instead of 3.11; which I assume is the same? So I'm not sure why that is hanging it up. I tried purging the pip cache, as well as installing with --ignore-requires-python; but since I think its using python itself to build a wheel with it it's just unhappy it has that .0 on the end.

Oddly, while I installed each of the dependencies individually trying to find which one was causing this numba issue, it dind't create this error again.

But, when installing RealTimeTTS[coqui]==0.4.1 I got a different error;
I got a mile long series of complaints related to a Cython compilation error.
```
... I didn't copy it all as it had a lot to say before this...
      [55/59] Cythonizing sklearn\utils\_vector_sentinel.pyx
      [56/59] Cythonizing sklearn\utils\_weight_vector.pyx
      [57/59] Cythonizing sklearn\utils\arrayfuncs.pyx
      [58/59] Cythonizing sklearn\utils\murmurhash.pyx
      [59/59] Cythonizing sklearn\utils\sparsefuncs_fast.pyx
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File ".venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 149, in prepare_metadata_for_build_wheel       
          return hook(metadata_directory, config_settings)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\setuptools\build_meta.py", line 174, in prepare_metadata_for_build_wheel
          self.run_setup()
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\setuptools\build_meta.py", line 268, in run_setup
          self).run_setup(setup_script=setup_script)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\setuptools\build_meta.py", line 158, in run_setup
          exec(compile(code, __file__, 'exec'), locals())
        File "setup.py", line 329, in <module>
          setup_package()
        File "setup.py", line 325, in setup_package
          setup(**metadata)
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\numpy\distutils\core.py", line 135, in setup
          config = configuration()
                   ^^^^^^^^^^^^^^^
        File "setup.py", line 187, in configuration
          config.add_subpackage("sklearn")
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\numpy\distutils\misc_util.py", line 1050, in add_subpackage
          config_list = self.get_subpackage(subpackage_name, subpackage_path,
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\numpy\distutils\misc_util.py", line 1016, in get_subpackage
          config = self._get_configuration_from_setup_py(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\numpy\distutils\misc_util.py", line 958, in _get_configuration_from_setup_py
          config = setup_module.configuration(*args)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "D:\_Offload_\TEMP\pip-install-4_5pc8l4\scikit-learn_51a2e99b9dde4bc2a9b62817d74103b9\sklearn\setup.py", line 85, in configuration
          cythonize_extensions(top_path, config)
        File "D:\_Offload_\TEMP\pip-install-4_5pc8l4\scikit-learn_51a2e99b9dde4bc2a9b62817d74103b9\sklearn\_build_utils\__init__.py", line 74, in cythonize_extensions  
          config.ext_modules = cythonize(
                               ^^^^^^^^^^
        File "D:\_Offload_\TEMP\pip-build-env-ohuxfq5e\overlay\Lib\site-packages\Cython\Build\Dependencies.py", line 1145, in cythonize
          result.get(99999)  # seconds
          ^^^^^^^^^^^^^^^^^
        File "Python311\Lib\multiprocessing\pool.py", line 774, in get
          raise self._value
      Cython.Compiler.Errors.CompileError: sklearn\ensemble\_hist_gradient_boosting\splitting.pyx
      [end of output]
```

Also, minor thing, should you have 'transformers' in the requirements.txt? I'm not sure what's best practices for stuff to be listed in it, but both times I have been prompted that I still needed it before I could try and run main.py

Anyway, with RealTimeTTS[coqui]==0.4.1 failing to install completely it is an unsupprising result the main.py fails to load. Stating RealtimeTTS isn't installed. I figured I'd post here at this point before I started going down the list installing each thing to see if I'm making a minor error on my end.

So far this is looking a lot better, and I'm not being facetious; I am genuinely happy to see new errors haha. I can look at this more later tomorrow and see if I can work it out more myself. But I really super appreciate you took the time to do this update, I may still be getting errors, but at least its not version conflicts this time!

Also, I didn't really consider it before; but in the requirements.txt you have opentelemetry and state its for fixing dependency problems? How so?

**Reply by kimjammer** [reply]
I haven't done a clean install in a bit, I should probably have transformers in the requirements, haha.
You might want to try using Python 3.11.9, which is the version I am using. Minor versions shouldn't have breaking changes, but you never know with building wheels and stuff - just might be worth trying. You also might want to see if you can install realtimeTTS by itself in a different venv.
The opentelemetry was because one of my dependencies needed something like version 1.0-1.9(made up numbers for example) and another needed 1.0-2.x and by default pip installed the newest version and broke the library that needed the older version. So it's just explicitly saying to use the older version that's compatible with both libraries.

**Reply by holdengand** [reply]
Python 3.11.9
Fresh venv
pip install realtimetts > Installs just fine, no issues.
pip install requirements.txt > still errors out on the numba thing.

pip install realtimetts
(removing RealTimeTTS[coqui]==0.4.1 from the requirements file)
pip install requirements.txt > completes fine
pip install transformers
pip install deepspeed-0.14.0+ce78a63-cp311-cp311-win_amd64.whl
python main.py
```
Starting Project...
Traceback (most recent call last):
  File "D:\_AI_\main.py", line 115, in <module>
    asyncio.run(main())
  File "Python311\Lib\asyncio\runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "Python\Python311\Lib\asyncio\runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "Python\Python311\Lib\asyncio\base_events.py", line 654, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "D:\_AI_\main.py", line 50, in main
    tts = TTS(signals)
          ^^^^^^^^^^^^
  File "D:\_AI_\tts.py", line 13, in __init__
    engine = CoquiEngine(
             ^^^^^^^^^^^^
TypeError: 'NoneType' object is not callable
```

And then I realized I forgot to install torch, torchvision and torchaudio; as directed on the readme. Torch 2.5.1 and torchaudio 2.5.1 had been installed by something in the process. Used your pip command in the readme, tried the main.py again and... same error. TypeError: 'NoneType' object is not callable.

I THINK I am past the instillation of dependencies, but I'm not sure what to do about TypeError: 'NoneType' object is not 
callable in relation to the CoquiEngine. It's the same error I had before. I attached a file of my current pip list for good measure.

But if its working for you then I am sure there must be something on my end I just am not understanding and simply messing up. This level of hands-on installation is about the most involved I have done in a while; and when it comes to my luck nothing like this 'just works' for me for some reason. I appreciate your efforts and I'm sorry all I have to offer is errors. Don't feel you need to rush to solve my minor problems!

[pipcurrent.txt](https://github.com/user-attachments/files/17710607/pipcurrent.txt)

**Reply by kimjammer** [reply]
Ah that's because if you install "realtimetts" by itself it doesn't install any of the engines (ie coqui engine is not installed) so it's NoneType when you try to use it. You have to specify realtimetts[coqui] to install the engine. Make sure you can install "realtimetts[coqui]==0.4.1" in an empty venv.

Don't worry about bothering me, I had to fight through my fair share of python library installation errors so I know how you feel.

**Reply by kimjammer** [reply, marked as answer]
If you can install realtimetts with coqui in another venv, try pulling my new commit and installing. I just went through a clean install and fixed all the dependency problems I ran into, so it should work. Most significantly, I now specify the numba and transformers version that's compatible with everything else.

**Reply by holdengand** [reply]
Oh right, that makes sense!

I took a nap and you seem to have fixed all my issues with the dependencies as far as I can tell. However I still can't get the main.py to run. 

I did not understand the errors I was getting, so I will provide the explanation Codeium is giving me for them.
```
The error message indicates that there is an issue with importing the log variable from the torch.distributed.elastic.agent.server.api module.

Looking at the provided code snippet from .aienv/Lib/site-packages/torch/distributed/elastic/agent/server/api.py, I see that log is not defined as a variable that can be imported. Instead, logger is defined as a variable that can be used for logging.

It seems that the deepspeed library is trying to import log from torch.distributed.elastic.agent.server.api, but it doesn't exist.

To fix this issue, you may need to update the deepspeed library to use the correct logging variable, which is logger. Alternatively, you can try to modify the torch.distributed.elastic.agent.server.api module to define log as an alias for logger.

However, without more information about your codebase and the specific requirements of your project, it's difficult to provide a more detailed solution.
```

I was able to resolve this issue;
I was first worried I was just not properly connecting to an LLM Endpoint and went ahead and resolved my issues with text-webui, and switching between it and kobold made no difference. ~But then suspected it was DeepSpeed again; I had assumed I was meant to be using that .14 wheel version that was produced by AllTalk. Double checked the freeze and saw it was the newer .15 version listed above. Pip installed that and that error was fixed.~ nope...

However I still have a series of issues related to Deepspeed and it thinking I am trying to use the CPU rather then GPU I think. I am not exactly sure where in this setup it is thinking that.  But there are also a few other errors. I had Codium produce a simplified list of the errors I was getting:
```
Error 1:
Error message: "This op had not been implemented on CPU backend."
File: deepspeed\ops\op_builder\cpu\no_impl.py
Line: 21
Related files:
deepspeed\model_implementations\transformers\ds_gpt.py (line 20)
deepspeed\module_inject\containers\gpt2.py (line 20)
deepspeed\module_inject\replace_module.py (line 385, 634, 694)
RealtimeTTS\engines\coqui_engine.py (line 501, 477)

Error 2:
Error message: "You are using torch.load with weights_only=False (the current default value), which uses the default pickle module implicitly."
File: TTS\tts\layers\xtts\xtts_manager.py
Line: 6
Related files:
TTS\utils\io.py (line 54)

Error 3:
Error message: "Config parameter replace_method is deprecated."
File: deepspeed\config_utils.py
Line: 70
Related files:
deepspeed\inference\engine.py (line 421)

Error 4:
Error message: "Config parameter mp_size is deprecated."
File: deepspeed\config_utils.py
Line: 70
Related files:
deepspeed\inference\engine.py (line 421)

Error 5:
Error message: "Setting accelerator to CPU. If you have GPU or other accelerator, we were unable to detect it."
File: real_accelerator.py
Line: 162
```

Anyway, I am quite pleased the dependencies all seem to work now! The requirements.txt is super straight forward now. ~At some point you might want to update the install instructions to direct to the newer deepspeed wheel though.~ I'm an idiot so ignore that.

I will look into this more on my own later this evening. If you happen to know what I am doing wrong though that would be nice. But at this point I am single-handedly going to be the source of all the troubleshooting getting this project started XD. Hopefully all these issues listed here will help anyone else at least for people who aren't developers or programmers, less familiar with these hand on installations like myself.

**Reply by holdengand** [reply]
Also, I just double checked to make sure my GPU was working within this venv;
nvcc --version
Cuda compilation tools, release 11.8, V11.8.89
Build cuda_11.8.r11.8/compiler.31833905_0

python -c "import torch; print(torch.cuda.is_available())"
False

Seems like its a CUDA issue? Related to torch?

**Reply by kimjammer** [reply]
Yup, I would double check that the torch version didn't get overriden. This might be a dumb question, but you have an NVIDIA gpu, right?

**Reply by kimjammer** [reply]
Not sure with what you mean with deepspeed, I'm still using 0.14.

**Reply by holdengand** [reply]
> This might be a dumb question, but you have an NVIDIA gpu, right?

This is fair. Yup, its an old Nvidia, but its a Nvidia. GTX 1070. I have used several other AI related projects before, and never really use CPU only.

> Not sure with what you mean with deepspeed, I'm still using 0.14.

I'm just an idiot and realize now I looked at one of my own pip freezes when I checked the deepspeed version. That was a last second check and I have installed so many times I am forgetting what's what. Sorry about that.

> I would double check that the torch version didn't get overriden.

Also, yes, I notice now this deepspeed wheel replaces the torch files. I fixed that and have made a note to pay more mind to that in the future.


My .env was getting a mismatch CUDA version error, specifically noting
```
RuntimeError: CUDA version mismatch! DeepSpeed ops were compiled and installed with a different version than what is being used at runtime. Please re-install DeepSpeed or switch torch versions. Install CUDA version=12.1, Runtime CUDA version=11.8
```

I thought this was an issue with what CUDA version I had activated in the venv while installing; but, bla bla bla, I did a bunch of stuff, long story short, I was using the 12.1 CUDA version of deepspeed not the 11.8 one.
`deepspeed-0.14.0+ce78a63-cp311-cp311-win_amd64.whl` ≠ `deepspeed-0.14.0+cu118-cp311-cp311-win_amd64.whl`

Anyway, installation complete. Hopefully my silly issues help someone else. Now its complaining about models not being in the right folder and I am confident I should be able to fix that myself after a bit of a break. So I haven't USED it yet, but I am past the dependency stuff. Thanks for your patience~

As of your version update, this should have been easy, so I marked it as the answer. Everything after was all my mistake in one form or another.

**Reply by kimjammer** [reply]
🎉 🎉 Congrats! Don't hesitate to ask if you have more questions.
